src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs

```diff
@@ -25,29 +24,73 @@ internal sealed partial class CSharpSelectionValidator(
     {
         private readonly bool _localFunction = localFunction;
 
-        protected override SelectionInfo UpdateSelectionInfo(
-            SelectionInfo selectionInfo, StatementSyntax? firstStatement, StatementSyntax? lastStatement, CancellationToken cancellationToken)
+        protected override InitialSelectionInfo GetInitialSelectionInfo(CancellationToken cancellationToken)
+        {
+            var root = this.SemanticDocument.Root;
+
+            var adjustedSpan = GetAdjustedSpan(OriginalSpan);
+
+            var firstTokenInSelection = root.FindTokenOnRightOfPosition(adjustedSpan.Start, includeSkipped: false);
+            var lastTokenInSelection = root.FindTokenOnLeftOfPosition(adjustedSpan.End, includeSkipped: false);
+
+            if (firstTokenInSelection.Kind() == SyntaxKind.None || lastTokenInSelection.Kind() == SyntaxKind.None)
+                return InitialSelectionInfo.Failure(FeaturesResources.Invalid_selection);
+
+            var commonRoot = firstTokenInSelection.GetCommonRoot(lastTokenInSelection);
+            var selectionInExpression = commonRoot is ExpressionSyntax;
+
+            var statusReason = CheckSpan();
+            if (statusReason is not null)
+                return InitialSelectionInfo.Failure(statusReason);
```

Comment:
simpler approach of just computing failure reason, then returning failure object in that one case, instead of instantiating failure object (verbosely) over and over again.